clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/QueryTool.java

```diff
@@ -80,55 +148,301 @@ public static Map<String, String> queryStoreForKey(
-        key = Boolean.parseBoolean(keyString);
-        break;
-      case STRING:
-        key = keyString;
-        break;
-      default:
+  public static Map<String, String> queryStoreWithCountByValue(
+      String store,
+      String keyString,
+      String url,
+      boolean isVsonStore,
+      Optional<String> sslConfigFile,
+      String countByValueFields,
+      int topK) throws Exception {
+
+    SSLFactory factory = null;
+    if (sslConfigFile.isPresent()) {
+      Properties sslProperties = SslUtils.loadSSLConfig(sslConfigFile.get());
+      String sslFactoryClassName = sslProperties.getProperty(SSL_FACTORY_CLASS_NAME, DEFAULT_SSL_FACTORY_CLASS_NAME);
+      factory = SslUtils.getSSLFactory(sslProperties, sslFactoryClassName);
+    }
+
+    if (url.toLowerCase().trim().startsWith("https") && (factory == null || factory.getSSLContext() == null)) {
+      throw new VeniceException("ERROR: The SSL configuration is not valid to send a request to " + url);
+    }
+
+    Map<String, String> outputMap = new LinkedHashMap<>();
+    try (AvroGenericStoreClient<Object, Object> client = ClientFactory.getAndStartGenericAvroClient(
+        ClientConfig.defaultGenericClientConfig(store)
+            .setVeniceURL(url)
+            .setVsonClient(isVsonStore)
+            .setSslFactory(factory))) {
+
+      // Parse keys
+      Set<Object> keys = parseKeys(keyString, client);
+
+      // Parse fields
+      String[] fields = countByValueFields.split(",");
+
+      // Create a pure client-side aggregation builder
+      ClientConfig clientConfig = ClientConfig.defaultGenericClientConfig(store)
+          .setVeniceURL(url)
+          .setVsonClient(isVsonStore)
+          .setSslFactory(factory);
+      AvroGenericReadComputeStoreClient<Object, Object> computeStoreClient =
+          (AvroGenericReadComputeStoreClient<Object, Object>) client;
+      AvroComputeAggregationRequestBuilder<Object> builder =
+          new AvroComputeAggregationRequestBuilder<>(computeStoreClient, ClientFactory.getSchemaReader(clientConfig));
+      builder.countGroupByValue(topK, fields);
+
+      ComputeAggregationResponse response = builder.execute(keys).get(60, TimeUnit.SECONDS);
+
+      outputMap.put("query-type", "countByValue");
+      outputMap.put("keys", keyString);
+      outputMap.put("fields", countByValueFields);
+      outputMap.put("topK", String.valueOf(topK));
+
+      // Add results for each field
+      for (String field: fields) {
+        Map<Object, Integer> valueCounts = response.getValueToCount(field);
+        outputMap.put(field + "-counts", valueCounts.toString());
+      }
+
+      return outputMap;
+    }
+  }
+
+  public static Map<String, String> queryStoreWithCountByBucket(
+      String store,
+      String keyString,
+      String url,
+      boolean isVsonStore,
+      Optional<String> sslConfigFile,
+      String countByBucketFields,
+      String bucketDefinitions) throws Exception {
+
+    SSLFactory factory = null;
```

Comment:
Maybe we could think about a way to use a private function avoid duplidate code.